### PR TITLE
ReturnExecutor

### PR DESCRIPTION
**Pull request by ObiWahn**
Adds the ReturnExecutor

### CODE REVIEW

**Comment by ObiWahn** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/2735/

File: arangod/Aql/ExecutionNode.cpp [hunk 1]
```diff
@@ -1836,7 +1835,31 @@ void ReturnNode::toVelocyPackHelper(VPackBuilder& nodes, unsigned flags) const {
...
     ExecutionEngine& engine, std::unordered_map<ExecutionNode*, ExecutionBlock*> const&) const {
-  return std::make_unique<ReturnBlock>(&engine, this);
+  ExecutionNode const* previousNode = getFirstDependency();
+  TRI_ASSERT(previousNode != nullptr);
+
+  auto it = getRegisterPlan()->varInfo.find(_inVariable->id);
+  TRI_ASSERT(it != getRegisterPlan()->varInfo.end());
+  RegisterId inputRegister = it->second.registerId;
+
+  //REMOVE this if register planning changes have been made and the ReturnExecutor is final
```
Comment (mchacki) [inline arangod/Aql/ExecutionNode.cpp]: What about this commented out LOGGING?
Is it useful in Production/Debugging ?
If yes please activate with a sensible log level, if not please delete.

File: arangod/Aql/ExecutorInfos.h [hunk 2]
```diff
@@ -65,6 +67,14 @@ class ExecutorInfos {
                 RegisterId nrInputRegisters, RegisterId nrOutputRegisters,
                 std::unordered_set<RegisterId> registersToClear);
 
+  //TODO - This constructor needs to be removed once the register planning is updated.
```
Comment (mchacki) [inline arangod/Aql/ExecutorInfos.h]: OK good, we keep this in mind!

File: arangod/Aql/ReturnExecutor.cpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+//   return it->second.registerId;
+// }
+
+ReturnExecutorInfos::ReturnExecutorInfos(RegisterId inputRegister, RegisterId outputRegister,
+                                         RegisterId nrInputRegisters, RegisterId nrOutputRegisters,
+                                         std::unordered_set<RegisterId> registersToClear,
+                                         bool returnInheritedResults)
+    : ExecutorInfos(make_shared_unordered_set({inputRegister}),
+                    make_shared_unordered_set({outputRegister}), nrInputRegisters,
+                    nrOutputRegisters, std::unordered_set<RegisterId>{} /*to clear*/,  // std::move(registersToClear),
```
Comment (mchacki) [inline arangod/Aql/ReturnExecutor.cpp]: please remove this  comment, makes it super hard to read / understand.

File: arangod/Aql/ReturnExecutor.cpp [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    return {state, std::move(stats)};
+  }
+
+  if (_infos._returnInheritedResults) {
+    output.copyRow(inputRow);
+  } else {
+    AqlValue val;
+    val = inputRow.getValue(_infos._inputRegisterId);
+    AqlValueGuard guard(val, true);
+    // LOG_DEVEL << "writing to ouputReg: " << _infos._outputRegisterId;
```
Comment (mchacki) [inline arangod/Aql/ReturnExecutor.cpp]: Please remove this comment

File: tests/Aql/ReturnExecutorTest.cpp [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+          std::tie(state, stats) = testee.produceRow(result);
+          REQUIRE(state == ExecutionState::DONE);
+          REQUIRE(!result.produced());
+        }
+      }
+    }
+  }
+
+  GIVEN("there are rows in the upstream") {
+    auto input = VPackParser::fromJson(
```
Comment (mchacki) [inline tests/Aql/ReturnExecutorTest.cpp]: Test is almost good.
Please Validate that the input is indeed returned!

Comment (ObiWahn) [reply tests/Aql/ReturnExecutorTest.cpp]: fixed

**Pull request merged**